lib/internal/fs/glob.js

```diff
@@ -169,212 +208,330 @@ class Glob {
+      if (stat) {
+        stat.name = firstPattern;
+        yield* this.addFromChildren(stat, path, fullpath, pattern, last, isLast);
+      }
+
       return;
     }
-    if (isFirst && pattern.at(0) === '.') {
-      // Start with ., proceed
-      this.#addSubpattern('.', pattern.child(new SafeSet([1])));
+
+    const children = this.#cache.readdirSync(fullpath);
+
+    for (const entry of children) {
+      yield* this.addFromChildren(entry, path, fullpath, pattern, last, isLast);
+    }
+  }
+
+  async *#addSubpatterns(path, pattern) {
+    const seen = this.#cache.add(path, pattern);
+    if (seen) {
       return;
     }
 
+    if (pattern.isFirst()) {
+      const shouldFinish = this.addPatternsOnFirst(pattern);
+
+      if (shouldFinish) {
+        return;
+      }
+    }
+
+    const fullpath = resolve(this.#root, path);
+    const stat = await this.#cache.stat(fullpath);
+    const last = pattern.last;
+    const isDirectory = stat?.isDirectory() || (stat?.isSymbolicLink() && pattern.hasSeenSymlinks);
+    const isLast = pattern.isLast(isDirectory);
+
+
     if (isLast && typeof pattern.at(-1) === 'string') {
       // Add result if it exists
       const p = pattern.at(-1);
-      const stat = this.#cache.statSync(join(fullpath, p));
+      const stat = await this.#cache.stat(join(fullpath, p));
       if (stat && (p || isDirectory)) {
-        ArrayPrototypePush(this.#results, join(path, p));
+        yield join(path, p);
       }
       if (pattern.indexes.size === 1 && pattern.indexes.has(last)) {
         return;
       }
     } else if (isLast && pattern.at(-1) === lazyMinimatch().GLOBSTAR &&
-     (path !== '.' || pattern.at(0) === '.' || (last === 0 && stat))) {
+      (path !== '.' || pattern.at(0) === '.' || (last === 0 && stat))) {
       // If pattern ends with **, add to results
       // if path is ".", add it only if pattern starts with "." or pattern is exactly "**"
-      ArrayPrototypePush(this.#results, path);
+      yield path;
     }
 
     if (!isDirectory) {
       return;
     }
 
-    let children;
     const firstPattern = pattern.indexes.size === 1 && pattern.at(pattern.indexes.values().next().value);
     if (typeof firstPattern === 'string') {
-      const stat = this.#cache.statSync(join(fullpath, firstPattern));
+      const stat = await this.#cache.stat(join(fullpath, firstPattern));
       if (stat) {
         stat.name = firstPattern;
-        children = [stat];
-      } else {
-        children = [];
+        yield* this.addFromChildren(stat, path, fullpath, pattern, last, isLast);
       }
-    } else {
-      children = this.#cache.readdirSync(fullpath);
+
+      return;
     }
 
-    for (let i = 0; i < children.length; i++) {
-      const entry = children[i];
-      const entryPath = join(path, entry.name);
-      this.#cache.addToStatCache(join(fullpath, entry.name), entry);
+    const children = await this.#cache.readdir(fullpath);
+
+    for (const entry of children) {
```

Comment:
dont use `for of`, since the prototype can be tampered

Reply:
can I use `for of` for going over `generator` or do I need to handle all the state by myself?